### PR TITLE
Report more informative CSS errors

### PR DESCRIPTION
**Pull request by jdm**
This requires https://github.com/servo/rust-cssparser/pull/143 for the final commit. There's no better way to split that work up, unfortunately, and it's extremely easy to bitrot. I would appreciate if we could expedite reviewing this work.

This is the work necessary to enable https://bugzilla.mozilla.org/show_bug.cgi?id=1352669. It makes sense to merge it separately because it's so much effort to keep it up to date with the ongoing Stylo work.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16752)

### CODE REVIEW

File: components/style/error_reporting.rs [hunk 1]
```diff
@@ -124,3 +124,15 @@ impl ParseErrorReporter for NullReporter {
...
+#[cfg(feature = "servo")]
+pub fn create_error_reporter() -> RustLogReporter {
+    RustLogReporter
+}
+
+/// Create an instance of the default error reporter for Stylo.
+#[cfg(feature = "gecko")]
+pub fn create_error_reporter() -> RustLogReporter {
+    RustLogReporter
+}
```
Comment (nox) [inline components/style/error_reporting.rs]: Aren't both of them the exact same function?

**Comment by Wafflespeanut** [comment]
cc @servo/style

File: components/style/error_reporting.rs (hunk 1 above)
Comment (jdm) [inline components/style/error_reporting.rs]: Yes. That changes in my patches to Gecko.

File: components/script_layout_interface/reporter.rs [hunk 2]
```diff
@@ -22,20 +22,20 @@ pub struct CSSErrorReporter {
...
+                        line_number_offset: u64) {
         let location = input.source_location(position);
         let line_offset = location.line + line_number_offset as usize;
         if log_enabled!(log::LogLevel::Info) {
             info!("Url:\t{}\n{}:{} {}",
                   url.as_str(),
                   line_offset,
                   location.column,
-                  message)
+                  error.to_string())
```
Comment (emilio) [inline components/script_layout_interface/reporter.rs]: Perhaps we should implement `Display` for `ContextualParseError`? Fine as a followup I guess.

File: components/selectors/parser.rs [hunk 3]
```diff
@@ -17,6 +18,17 @@ use std::slice;
 use tree::SELECTOR_WHITESPACE;
 use visitor::SelectorVisitor;
 
+#[derive(Clone, Debug, PartialEq)]
+pub enum SelectorParseError<T> {
+    PseudoElementInComplexSelector,
+    NoQualifiedNameInAttributeSelector,
+    TooManyCompoundSelectorComponents,
```
Comment (emilio) [inline components/selectors/parser.rs]: Let's rename this to `TooManyCompoundSelectorComponentsInNegation`, or perhaps `TooComplexNegationSelector`?

File: components/style/error_reporting.rs [hunk 4]
```diff
@@ -6,22 +6,150 @@
...
+    InvalidCounterStyleWithoutAdditiveSymbols,
+    /// A counter style rule had extends with symbols.
+    InvalidCounterStyleExtendsWithSymbols,
+    /// A counter style rule had extends with additive-symbols.
+    InvalidCounterStyleExtendsWithAdditiveSymbols
+}
+
+impl<'a> ContextualParseError<'a> {
+    /// Turn a parse error into a string representation.
+    pub fn to_string(&self) -> String {
```
Comment (emilio) [inline components/style/error_reporting.rs]: Perhaps we should implement `Display` and implement `to_string` as `format!("{}", self)`? Fine also as a followup/easy issue I guess :)

File: components/style/properties/longhand/text.mako.rs [hunk 5]
```diff
@@ -175,34 +177,40 @@ ${helpers.single_keyword("unicode-bidi",
...
-                        "line-through" => if result.contains(LINE_THROUGH) { return Err(()) }
-                                          else { empty = false; result.insert(LINE_THROUGH) },
-                        "blink" => if result.contains(BLINK) { return Err(()) }
-                                   else { empty = false; result.insert(BLINK) },
-                        _ => return Err(())
+        loop {
+            let result: Result<_, ParseError> = input.try(|input| {
+                match input.expect_ident() {
+                    Ok(ident) => {
+                        (match_ignore_ascii_case! { &ident,
```
Comment (emilio) [inline components/style/properties/longhand/text.mako.rs]: I think it shouldn't be needed to wrap it in parenthesis anymore?

File: components/style/values/generics/position.rs [hunk 6]
```diff
@@ -104,23 +104,23 @@ impl<L: HasViewportPercentage> HasViewportPercentage for PositionValue<L> {
...
+    pub fn parse_internal<'i, 't, F>(context: &ParserContext, input: &mut Parser<'i, 't>,
+                             mut is_allowed_keyword: F) -> Result<PositionValue<L>, ParseError<'i>>
         where F: FnMut(Keyword) -> bool
     {
         let (mut pos, mut keyword) = (None, None);
         for _ in 0..2 {
             if let Ok(l) = input.try(|i| L::parse(context, i)) {
                 if pos.is_some() {
-                    return Err(())
+                    return Err(StyleParseError::UnspecifiedError.into())
```
Comment (emilio) [inline components/style/values/generics/position.rs]: Is there an intention to convert all the `UnspecifiedError`s into more explicit error kinds as followup work?

File: components/style/values/mod.rs [hunk 7]
```diff
@@ -237,13 +243,16 @@ pub struct CustomIdent(pub Atom);
...
-        match_ignore_ascii_case! { &ident,
-            "initial" | "inherit" | "unset" | "default" => return Err(()),
-            _ => {}
+    pub fn from_ident<'i>(ident: Cow<'i, str>, excluding: &[&str]) -> Result<Self, ParseError<'i>> {
+        let valid = match_ignore_ascii_case! { &ident,
+            "initial" | "inherit" | "unset" | "default" => false,
+            _ => true
         };
+        if !valid {
+            return Err(BasicParseError::UnexpectedToken(Token::Ident(ident)).into());
```
Comment (emilio) [inline components/style/values/mod.rs]: Perhaps just use braces in the upper `match` branch and get rid of `valid`?

File: components/style/values/specified/align.rs [hunk 8]
```diff
@@ -340,77 +340,78 @@ impl Parse for JustifyItems {
...
 
     let ident = input.expect_ident()?;
-    match_ignore_ascii_case! { &ident,
+    (match_ignore_ascii_case! { &ident,
         "auto" => Ok(ALIGN_AUTO),
         "normal" => Ok(ALIGN_NORMAL),
         "stretch" => Ok(ALIGN_STRETCH),
         _ => Err(())
-    }
+    }).map_err(|()| BasicParseError::UnexpectedToken(Token::Ident(ident)).into())
```
Comment (emilio) [inline components/style/values/specified/align.rs]: I'm assuming this funky way of doing this is because `ident` is borrowed in the `match` expression, is that right? If so, that's kind of unfortunate, I guess... :(

File: components/style/values/specified/basic_shape.rs [hunk 9]
```diff
@@ -120,11 +123,12 @@ impl InsetRect {
...
 
 impl Parse for InsetRect {
-    fn parse(context: &ParserContext, input: &mut Parser) -> Result<Self, ()> {
+    fn parse<'i, 't>(context: &ParserContext, input: &mut Parser<'i, 't>)
+                     -> Result<Self, ParseError<'i>> {
         match input.try(|i| i.expect_function()) {
             Ok(ref s) if s.eq_ignore_ascii_case("inset") =>
                 input.parse_nested_block(|i| GenericInsetRect::parse_function_arguments(context, i)),
-            _ => Err(())
+            _ => Err(BasicParseError::ExpectedToken(Token::Function("inset".into())).into())
```
Comment (emilio) [inline components/style/values/specified/basic_shape.rs]: Do you know if this `"inset".into()` will allocate or not? I'm assuming that since it's an static string it shouldn't need to, but I worry that perhaps to coerce it to do that you may need to be explicit about it. Not a big deal in any case I think

**Review by emilio — COMMENTED** [review]
Wow, that's a big diff. Haven't gone through it with detail, would love help from @nox/@simonsapin reviewing.

Some questions below.

File: components/style/values/mod.rs (hunk 7 above)
Comment (emilio) [inline components/style/values/mod.rs]: nvm about this one.

File: components/style/values/generics/position.rs (hunk 6 above)
Comment (jdm) [inline components/style/values/generics/position.rs]: Yes. It's a nice source of easy issues and a lot of effort for this patch in particular.

File: components/style/values/specified/align.rs (hunk 8 above)
Comment (jdm) [inline components/style/values/specified/align.rs]: Yep! Best compromise I could find.

File: components/style/values/specified/basic_shape.rs (hunk 9 above)
Comment (jdm) [inline components/style/values/specified/basic_shape.rs]: It should be a Cow::Borrowed afaik.

File: components/style/properties/longhand/text.mako.rs (hunk 5 above)
Comment (jdm) [inline components/style/properties/longhand/text.mako.rs]: Fixing this will be kind of annoying given how often this pattern appears. I would prefer not to, if that's alright.

Comment (emilio) [inline components/style/properties/longhand/text.mako.rs]: Yeah, that's totally fine for me :)

**Comment by SimonSapin** [comment]
Reviewed 9 of 9 files at r1, 4 of 4 files at r2, 1 of 1 files at r3, 75 of 75 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[components/style/custom_properties.rs, line 241 at r4](https://reviewable.io:443/reviews/servo/servo/16752#-KkRWV_K5ERWndP1yyvY:-KkRWV_K5ERWndP1yyvZ:bsfkc9a) ([raw file](https://github.com/servo/servo/blob/8d26cb8ebafd19e79cf41bb3388087ffe4e4f4f2/components/style/custom_properties.rs#L241)):*
> ```Rust
>                 return Err(StyleParseError::CloseSquareBracketInDeclarationValueBlock.into()),
>             Token::CloseCurlyBracket =>
>                 return Err(StyleParseError::CloseCurlyBracketInDeclarationValueBlock.into()),
> ```

These are not just (various kinds of) closing brackets, they are unbalanced / improperly nested ones.

---

*[components/style/media_queries.rs, line 227 at r4](https://reviewable.io:443/reviews/servo/servo/16752#-KkRY-GyzL3KtiiFXlWG:-KkRY-GyzL3KtiiFXlWH:b-c4e3ny) ([raw file](https://github.com/servo/servo/blob/8d26cb8ebafd19e79cf41bb3388087ffe4e4f4f2/components/style/media_queries.rs#L227)):*
> ```Rust
>             Ok(ident) => {
>                 let result: Result<_, ParseError> = MediaQueryType::parse(&*ident)
>                     .map_err(|()| BasicParseError::UnexpectedToken(Token::Ident(ident)).into());
> ```

Could you avoid creating tokens like this? I have plans to change the `Token` data structure (to avoid more memory allocations) that involve structs with private fields, so that only the tokenizer will be able to create tokens. Maybe add a `UnexpectedIdent` error variant instead?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/16752)*

**Comment by jdm** [comment]
@SimonSapin I have addressed the review comments and made a new commit which applies to the existing code that you reviewed. I'm going to go ahead and rebase all of my changes separately now, and I won't push them until you are done reviewing the existing code.

**Comment by SimonSapin** [comment]
r+

---

Reviewed 54 of 54 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[components/style/values/specified/basic_shape.rs, line 131 at r4](https://reviewable.io:443/reviews/servo/servo/16752#-KjQtC6mKAgMQrEv7NEc-r4-131:-Km3MDxFZQRnJlutud4F:b-qvw9aj) ([raw file](https://github.com/servo/servo/blob/8d26cb8ebafd19e79cf41bb3388087ffe4e4f4f2/components/style/values/specified/basic_shape.rs#L131)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary><blockquote>

It should be a Cow::Borrowed afaik.
</blockquote></details>

Indeed, this uses `Cow::Borrowed` and does not allocate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/16752#-:-Km3N6yN1qB_teoNHFOn:b32wu6)*

**Comment by nox** [comment]
@jdm What was the problem with not sharing the same `ParseError` across all impls again?

File: components/style/properties/longhand/background.mako.rs [hunk 10]
```diff
@@ -128,17 +128,20 @@ ${helpers.predefined_type("background-image", "ImageLayer",
...
-            }
-        }
+            _ => Err(()),
+        }).or_else(|()| {
+            let horizontal: Result<_, ParseError> = RepeatKeyword::from_ident(&ident)
+                .map_err(|()| SelectorParseError::UnexpectedIdent(ident).into());
+            let horizontal = try!(horizontal);
+            let vertical = input.try(RepeatKeyword::parse).ok();
+            Ok(SpecifiedValue::Other(horizontal, vertical))
+        })
```
Comment (nox) [inline components/style/properties/longhand/background.mako.rs]: @jdm Why was that change needed?

Comment (nox) [inline components/style/properties/longhand/background.mako.rs]: Oh yeah borrowing issues, unfortunate. :(

**Comment by jdm** [comment]
@nox I'm not sure I understand the question. If it is why rust-cssparser's ParseError does not contain every possible error, the answer is because that's really inconvenient API design.
